src/transformers/models/llama/modeling_llama.py

```diff
@@ -1216,11 +1260,36 @@ def prepare_inputs_for_generation(
             if past_key_values:
                 position_ids = position_ids[:, -input_ids.shape[1] :]
 
+                # This `clone` call is needed to avoid recapturing cuda graphs with `torch.compile`'s  `mode="reduce-overhead`, as otherwise the input `position_ids` would have various stride during the decoding. Here, simply using `.contiguous()` is not sufficient as in the batch size = 1 case, `position_ids` is already contiguous but with varying stride which retriggers a capture.
+                position_ids = position_ids.clone(memory_format=torch.contiguous_format)
+
         # if `inputs_embeds` are passed, we only want to use them in the 1st generation step
         if inputs_embeds is not None and cache_position[0] == 0:
             model_inputs = {"inputs_embeds": inputs_embeds}
         else:
-            model_inputs = {"input_ids": input_ids.contiguous()}  # `contiguous()` needed for compilation use cases
+            model_inputs = {"input_ids": input_ids}
+
+        if isinstance(past_key_values, StaticCache) and attention_mask.ndim == 2:
```

Comment:
We need a TODO for me/Arthur here :D 

TL;DR `StaticCache` is not the only compileable cache. At the moment we don't have an attribute that says "this cache supports compilation of the fwd pass". We should create that attribute and use it here

Reply:
+1 on this! Currently, our check for Gemma-2 is looking at:
```python
       if isinstance(past_key_values, HybridCache) and attention_mask.ndim == 2:
```

=> it would be great to remove the ambiguity and truly make this a cache agnostic function

Reply:
I think we should create the attention mask once indeed, and just "update" it to unpad after each decoding step (pay attention) , if possible in generate, this way we don't change prepare_inputs_for_generation. We can do better checks of the Cache class in generate as well